cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,538 @@
+  }
+
+  @Nullable
+  private Replica leaderReplicaOfTopicPartition(ClusterModel clusterModel, TopicPartition topicPartition) {
+    for (Replica leaderReplica : clusterModel.leaderReplicas()) {
+      if (leaderReplica.topicPartition().equals(topicPartition)) {
+        return leaderReplica;
+      }
+    }
+    return null;
+  }
+
+
+  private Set<Broker> eligibleBrokersForLeadership(ClusterModel clusterModel, OptimizationOptions optimizationOptions) {
+    return clusterModel.brokers()
+                       .stream()
+                       .filter(broker -> isEligibleToHaveLeaders(broker, optimizationOptions))
+                       .collect(Collectors.toSet());
+  }
+
+  private boolean isEligibleToHaveLeaders(Broker broker, OptimizationOptions optimizationOptions) {
+    return broker.isAlive()
+           && !optimizationOptions.excludedBrokersForLeadership().contains(broker.id())
+           && !optimizationOptions.excludedBrokersForReplicaMove().contains(broker.id());
+  }
+
+  private void moveAwayOfflineReplicas(Broker srcBroker,
+                                       ClusterModel clusterModel,
+                                       Set<Goal> optimizedGoals,
+                                       OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    if (srcBroker.currentOfflineReplicas().isEmpty()) {
+      return;
+    }
+    SortedSet<Broker> eligibleBrokersToMoveOfflineReplicasTo = findEligibleBrokersToMoveOfflineReplicasTo(clusterModel, optimizationOptions);
+    eligibleBrokersToMoveOfflineReplicasTo.removeIf(broker -> broker.id() == srcBroker.id());
+    if (eligibleBrokersToMoveOfflineReplicasTo.isEmpty()) {
+      throw new OptimizationFailureException("Cannot move all replicas away from broker " + srcBroker);
+    }
+    Set<Replica> offlineReplicas = new HashSet<>(srcBroker.currentOfflineReplicas());
+    for (Replica offlineReplica : offlineReplicas) {
+      if (maybeApplyBalancingAction(clusterModel, offlineReplica, eligibleBrokersToMoveOfflineReplicasTo,
+                                    INTER_BROKER_REPLICA_MOVEMENT, optimizedGoals, optimizationOptions) == null) {
+        throw new OptimizationFailureException(String.format("Cannot move offline replica %s to any broker in %s",
+                                                             offlineReplica, eligibleBrokersToMoveOfflineReplicasTo));
+      }
+    }
+  }
+
+  private SortedSet<Broker> findEligibleBrokersToMoveOfflineReplicasTo(ClusterModel clusterModel, OptimizationOptions optimizationOptions) {
+    Set<Broker> candidateBrokers = clusterModel.brokers()
+                                               .stream()
+                                               .filter(broker -> broker.isAlive()
+                                                                 && !optimizationOptions.excludedBrokersForReplicaMove().contains(broker.id()))
+                                               .collect(Collectors.toSet());
```

Comment:
This whole filtering is redundant. The return value of this function is used in `maybeApplyBalancingAction`. But if we check the content of this function, we will notice that (1) it does the same filtering for excluded brokers for both leadership and replica moves and (2) it does not move a replica to a broker that has the same replica, because it is not a `legitMove`.

Can we 
1. just use:
```
    SortedSet<Broker> sortedBrokers = new TreeSet<>(
        Comparator.comparingInt((Broker broker) -> broker.replicas().size()).thenComparingInt(Broker::id));
    sortedBrokers.addAll(clusterModel.aliveBrokers());
```
2. delete `findEligibleBrokersToMoveOfflineReplicasTo ` and move its content to `moveAwayOfflineReplicas`?

Reply:
Good point